src/mscorlib/shared/System/String.cs

```diff
@@ -2,109 +2,391 @@
-        // than 0x80) before security is fully initialized.  Without security initialized, we can't grab resources (the nlp's)
-        // from the assembly.  This provides a workaround for that problem and should NOT be used anywhere else.
-        //
-        internal unsafe static string SmallCharToUpper(string strIn)
-        {
-            Debug.Assert(strIn != null);
-            //
-            // Get the length and pointers to each of the buffers.  Walk the length
-            // of the string and copy the characters from the inBuffer to the outBuffer,
-            // capitalizing it if necessary.  We assert that all of our characters are
-            // less than 0x80.
-            //
-            int length = strIn.Length;
-            String strOut = FastAllocateString(length);
-            fixed (char* inBuff = &strIn._firstChar, outBuff = &strOut._firstChar)
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        public extern String(char[] value, int startIndex, int length);
+
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private string Ctor(char[] value, int startIndex, int length)
+        {
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+
+            if (startIndex < 0)
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndex);
+
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length), SR.ArgumentOutOfRange_NegativeLength);
+
+            if (startIndex > value.Length - length)
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
+
+            if (length > 0)
             {
-                for (int i = 0; i < length; i++)
+                string result = FastAllocateString(length);
+                unsafe
                 {
-                    int c = inBuff[i];
-                    Debug.Assert(c <= 0x7F, "string has to be ASCII");
+                    fixed (char* dest = &result._firstChar, source = value)
+                    {
+                        wstrcpy(dest, source + startIndex, length);
+                    }
+                }
+                return result;
+            }
+            else
+                return Empty;
+        }
+
+        [CLSCompliant(false)]
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        unsafe public extern String(char* value);
+
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private unsafe string Ctor(char* ptr)
+        {
+            if (ptr == null)
+                return Empty;
+
+            int count = wcslen(ptr);
+            if (count == 0)
+                return Empty;
+
+            string result = FastAllocateString(count);
+            fixed (char* dest = &result._firstChar)
+                wstrcpy(dest, ptr, count);
```

Comment:
Nit: in some places the wstrcpy is wrapped in braces, in some it's not.  It'd be nice to be consistent.